### PR TITLE
Cross-project flows

### PR DESCRIPTION
**Pull request by davisagli**
This adds the ability for a project to include other projects from external sources (so far just github) and make use of tasks and flows from the included project.

## Example: Creating an org with NPSP from the V4S repo

Include NPSP in V4S's cumulusci.yml:

    sources:
        npsp:
            github: https://github.com/SalesforceFoundation/Cumulus

Now run `cci flow run npsp:install_beta`. You should see a line saying the Cumulus source was fetched, and end up with an org with NPSP including the full trial config.

## Details

A github source can specify a specific `commit`, `tag`, or `branch`. If none is specified, CumulusCI will try to find the tag from the latest release. If there's no release, it'll use master.

The external source is fetched into `.cci/projects/[repo name]/[commit hash]`. This means that repeated runs do not need to re-fetch the source each time if the commit is the same. (However this cache is not currently shared between multiple CumulusCI projects; maybe it should be.)

Task configs and Flow configs now get a project_config attached to them so we can keep track of which project they came from. Tasks now explicitly change directories to their project's repository root while executing, so that all our tasks which were written assuming the current directory is the repo root still work.

https://github.com/SFDO-Tooling/MetaDeploy/pull/467 is the corresponding PR for MetaDeploy that will be merged after this to add support for running steps from different sources there.

### CODE REVIEW

**Comment by prescod** [comment]
In my testing, if there was no release and no commit, github.py throws an exception

https://api.github.com/repos/SFDO-Tooling/core-sfdo-performance-tests/releases/latest

```
Traceback (most recent call last):
  File "/Users/pprescod/code/envs/cumulusci-py3/bin/cci", line 11, in <module>
    load_entry_point('cumulusci', 'console_scripts', 'cci')()
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/pprescod/code/CumulusCI/cumulusci/cli/cci.py", line 208, in new_func
    func(config, *args, **kw)
  File "/Users/pprescod/code/CumulusCI/cumulusci/cli/cci.py", line 1336, in flow_run
    handle_exception_debug(config, debug, no_prompt=no_prompt)
  File "/Users/pprescod/code/CumulusCI/cumulusci/cli/cci.py", line 1327, in flow_run
    coordinator = config.get_flow(flow_name, options=options)
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/runtime.py", line 96, in get_flow
    flow_config = self.project_config.get_flow(name)
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/config/BaseProjectConfig.py", line 753, in get_flow
    other_config = self.get_namespace(ns)
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/config/BaseProjectConfig.py", line 769, in get_namespace
    project_config = self.include_source(spec)
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/config/BaseProjectConfig.py", line 775, in include_source
    source = GitHubSource(self, spec)
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/config/BaseProjectConfig.py", line 804, in __init__
    self.resolve()
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/config/BaseProjectConfig.py", line 846, in resolve
    release = _find_latest_release(self.repo, include_beta=False)
  File "/Users/pprescod/code/CumulusCI/cumulusci/core/config/BaseProjectConfig.py", line 884, in _find_latest_release
    return repo.latest_release()
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/github3/repos/repo.py", line 2104, in latest_release
    json = self._json(self._get(url), 200)
  File "/Users/pprescod/code/envs/cumulusci-py3/lib/python3.7/site-packages/github3/models.py", line 156, in _json
    raise exceptions.error_for(response)
github3.exceptions.NotFoundError: 404 Not Found

```

**Comment by prescod** [comment]
After I added a branch, it seemed to work well for my use-case!

**Comment by davisagli** [comment]
@prescod Thanks for trying it. I added a fix for that case.

File: cumulusci/core/config/project_config.py [hunk 1]
```diff
@@ -727,11 +733,53 @@ def process_github_dependency(  # noqa: C901
...
+            flow_config.name = name
+            flow_config.project_config = other_config
+        else:
+            flow_config = super().get_flow(name)
+            flow_config.name = name
+            flow_config.project_config = self
+        return flow_config
+
+    def get_namespace(self, ns):
+        spec = getattr(self, "sources__{}".format(ns))
```
Comment (Br4nd0R) [inline cumulusci/core/config/project_config.py]:
```suggestion
        spec = getattr(self, f"sources__{ns}")
```

File: cumulusci/core/config/project_config.py [hunk 2]
```diff
@@ -727,11 +733,53 @@ def process_github_dependency(  # noqa: C901
...
+        if spec is None:
+            raise NamespaceNotFoundError("Namespace not found: {}".format(ns))
```
Comment (Br4nd0R) [inline cumulusci/core/config/project_config.py]:
```suggestion
            raise NamespaceNotFoundError(f"Namespace not found: {ns}")
```

File: cumulusci/core/config/project_config.py [hunk 3]
```diff
@@ -727,11 +733,53 @@ def process_github_dependency(  # noqa: C901
...
+        else:
+            if "github" in spec:
+                source = GitHubSource(self, spec)
+            elif "path" in spec:
+                source = LocalFolderSource(self, spec)
             else:
-                return repo.latest_release()
-        except Exception as e:
-            self.logger.warning("{}: {}".format(e.__class__.__name__, str(e)))
+                raise Exception("Not sure how to load project: {}".format(spec))
```
Comment (Br4nd0R) [inline cumulusci/core/config/project_config.py]:
```suggestion
                raise Exception(f"Not sure how to load project: {spec}")
```

File: cumulusci/core/config/project_config.py [hunk 4]
```diff
@@ -727,11 +733,53 @@ def process_github_dependency(  # noqa: C901
 
         return repo_dependencies
 
-    def _find_latest_release(self, repo, include_beta=None):
-        try:
-            if include_beta:
-                return next(repo.releases())
+    def get_task(self, name):
```
Comment (Br4nd0R) [inline cumulusci/core/config/project_config.py]: Some doc strings for `get_task` and other methods would be good.

File: cumulusci/core/source/github.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            if release:
+                ref = "tags/" + release.tag_name
+            else:
+                ref = "heads/" + self.repo.default_branch
+        self.description = ref[6:] if ref.startswith("heads/") else ref
+        self.commit = self.repo.ref(ref).object.sha
+
+    def fetch(self, path=None):
+        """Fetch the archive of the specified commit and construct its project config."""
+        # To do: copy this from a shared cache
```
Comment (Br4nd0R) [inline cumulusci/core/source/github.py]: did you want to address this TODO on this iteration of the functionality?

Comment (davisagli) [reply cumulusci/core/source/github.py]: I'm intentionally leaving this one for later. It's a nice-to-have for reducing network traffic a bit, but isn't critical and shouldn't change functionality.

File: cumulusci/core/config/project_config.py (hunk 4 above)
Comment (davisagli) [reply cumulusci/core/config/project_config.py]: Added!

**Pull request merged**